# Significant differences between CWL and WDL versions
This list is not intended as an exhaustive list of every difference. These are two different workflow languages using two different inline languages to generate configuration files after all! Instead, the purpose of this document is to keep track of changes which may have implications for users or maintainers of this code.  
# All Workflows 
## Strictly Necessary Changes  
* Filenames are generated by calling the Seven Bridges API in the CWL. This isn't possible in WDL, so inputs are generated from JSONs instead.  
* The original CWL does not have an option for disk space, but the WDL does, as it is a soft requirement for running on Terra.  
* The CWL generates config files using an InlineJavascriptRequirement, which is run before the CWL equivivalent of a task's command section begins. The WDL generates them using an inline Python script during the beginning of a task's command section.  

## "Not Strictly Necessary But In WDL This Makes More Sense"
* The original CWL allocates memory and CPUs on an overall level, while the WDL does it on a task level. In other words, the WDL is more granular.  
	* Reasoning: GCS has stricter requirements than AWS regarding storage. Doing it this way also allows resource-heavy tasks (such as check_gds in vcf-to-gds-wf) to only use the large number of resources they need in their respective task, instead of the whole workflow.  

## Different Design Choices
* The CWL imports other CWLs. The WDL does not import other WDLs, except in the case of the checker workflow.  
	* Reasoning: It is possible for WDLs to contain just tasks and be imported into another task, but in some contexts it can be slightly less secure.

# Workflow A --- vcf-to-gds-wf.wdl
## Strictly Necessary Changes  
* The WDL relocalizes all files into the working directory in the unique variant IDs step. [This issue is explained in detail here](https://github.com/DataBiosphere/analysis_pipeline_WDL/issues/2).
* In order to pass chromosome numbers into check_gds, chromosome number is written to a file (completely seperate from the configuration file) in the inline Python section. Upon exiting the Python block, this extra file is read in BASH and then passed to the Rscript call as an argument (as opposed to being in the configuration file). Although the actual call to the R script is identical as the CWL also passes the chr number as an argument, the CWL is able to rely on its inline Javascript to do this, while the WDL must use this bizarre workaround to pass the chr number out of the Python scope.     

## "Not Strictly Necessary But In WDL This Makes More Sense"
* The WDL will not start the check_gds task if check_gds is false. The CWL will start the check_gds task regardless and generate a config file, and the true/false only applies to calling the  script.
	* Reasoning: The way GCS billing works, this has the potential of being cheaper. Otherwise we would spend for having a powerful non-preemptible compute designed for an intense task, then only using that compute for making a text file.

## Error Handling
* The WDL can correctly handle a mixture of file extensions, not so much by design, but due to the specifics of implementation. The CWL will handle such a mixture incorrectly in check_gds (but can correctly handle a homogenous group, such as all being bcf files).

# Workflow B --- ld-pruning-wf.wdl
## Strictly Necessary Changes
* Similiar to Workflow A's file relocalization trick in unique_variant_ids, this workflow's merge_gds task has to use the same workaround.
* WDL does not have an equivalent to ScatterMethod:DotProduct so it instead scatters using zip().
* The workaround used by Workflow A's check_gds to get chromosome number (see above) is also used by check_merged_gds for the same reason.  

## Different Design Choices
* The output prefix is not used for the subset_gds task in the CWL, but is used in the WDL. It appears to be intended to be used based on what I see in the CWL's code, but on SB it does not get used in that task.  

## Error Handling
* The CWL appears to contain a bug where `exclude_PCA_cor` being set to false is not respected. See #14 for more info. The WDL avoids this using slightly different logic.

